url587829ef81_champdavis_KhovanovHomology_tgz-pJ8-KhovanovMapJ8/src/Knot.java

```diff
@@ -0,0 +1,227 @@
+                this.vertices[this.smoothing.getFour()-1].setFrom(current);
+            }
+            
+            else{
+                current.setTo(this.vertices[this.smoothing.getThree()-1]);
+                this.vertices[this.smoothing.getThree()-1].setFrom(current);
+            }
+            
+            current.setFrom(prev);
+              
+        }
+        
+        //Fix Crossings
+        int flag;
+        for(Crossing cross: this.crossings){
+            flag=0;
+            if(this.vertices[cross.get1st()-1].getTo().getNum()!=cross.get2nd()){
+                cross.swapOverstrand();
+                flag++;
+            }
+            if(this.vertices[cross.get3rd()-1].getTo().getNum()!=cross.get4th()){
+                cross.swapUnderstrand();
+                flag++;
+            }
+            if(flag==1){
+                cross.swapSign();  
+            }       
+        }
+        
+        /*System.out.println(this.binarycode+"     "+this.smoothing.toString());
+        for(int i=0; i<v.length; i++){
+            System.out.println(this.vertices[i].toString());
+        }*/
+        
+        makeCircles();
+        
+        
+        /*for(Vertex vert:this.vertices){
+            System.out.println(vert.toString());
+            
+        }*/
+    }
+    
+    public int numCrossings(){ return this.crossings.length;}
+    public String getBinCode(){ return this.binarycode;}
+    public Knot getLeftKnot(){ return this.leftknot;}
+    public Knot getRightKnot(){ return this.rightknot;}
+        
+    
+    public void resolveknot(){
+        if(this.crossings.length==0){
+            return;
+        }
+        else{
+            //what crossing we are resolving
+            Crossing current = this.crossings[0];
+            //System.out.println("CURRENT CROSSING: "+current.toString());
+            Crossing[] newcrossings = new Crossing[this.crossings.length-1];   //make next level's crossing array; one less because we resolved a crossing
+            for(int i=1; i<crossings.length; i++){                                //fill up the new array
+                newcrossings[i-1] = this.crossings[i];
+            }            
+            
+            this.leftknot=new Knot(newcrossings, this.binarycode+"0", current.getZeroSmoothing(), this.vertices);
+            this.rightknot=new Knot(newcrossings, this.binarycode+"1", current.getOneSmoothing(), this.vertices);
+            this.leftknot.resolveknot();
+            this.rightknot.resolveknot();
+        }
+    }
+    
+    public void makeCircles(){
+        String code;
+        Vertex current;
+        for(Vertex vert: this.vertices){
+            code = "";
+            current = vert;
+            while(!current.wasVisited()){
+                code+=(""+current.getNum());
+                current.visit();
+                current=current.getTo();
+                
+            }
+            //System.out.println(code);
+            if(code!=""){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.